custom-schemes.md

```diff
@@ -0,0 +1,41 @@
+
+2. Can wallets limit requests to secure contexts?
+
+   - If a user uses a website via http (not https) and it invokes a custom scheme, it's possible that an attacker with access to the network could eavesdrop on or even modify the request (such as by replacing the reader public key with their own). 
+   - A browser API could take the position that it must function only within the context of a secure context (https).
+
+3. What is the user experience when multiple wallet apps are installed?
+
+   - When multiple Android applications register for the same custom URL scheme the operating system presents the user with a picker every time it is invoked. Even if the verifier provided the user with appropriate context, a user may have trouble remembering which application has which credentials and so be more likely to cancel or fail the request.
+   - On iOS when multiple applications are installed that register for the same custom URL scheme, the behavior is [undefined](https://stackoverflow.com/questions/13130442/multiple-apps-with-the-same-url-scheme-ios). One wallet will get invoked without any indication that other wallets were available.
+
+4. What is the user experience when no wallet app is installed or when the user cancels wallet invocation?
+
+   - Browsers generally silently ignore navigations to URL schemes with no handlers installed. This likely means a website cannot distinguish between the time it takes for a wallet to process a request and nothing happening due to the lack of any wallet app being installed. 
+   - Additionally browsers on Android and iOS often offer the user the choice of whether or not to invoke an application for a custom URL scheme navigation (even if there is only one such app). The user canceling at this stage is also indistinguishable from no valid application being installed.
+   - A browser-based API could show the user some meaningful error message when no relevant wallets are installed.
+
+5. What is the user experience on a desktop operating system?
+
+   - Most wallet applications exist only as mobile applications, and users are generally reluctant to install native applications on desktop operating systems. What fallback will flows relying on custom URL schemes provide for users of desktop operating systems?
+   - A browser-based API can potentially provide low-friction integration between a desktop browser and a wallet application running on a mobile phone.
+
+6. What are the privacy implications of a wallet accepting custom schemes?
+
+   - In general Android works to keep the list of applications installed private from other applications. However applications registering to handle custom URL schemes effectively make their package name visible to other applications. By registering to support custom URL schemes a wallet application may be leaking PII (such as the user's likely citizenship status or state of residency) to any other application on the device.
```

Comment:
The mDL wallets that implement  these schemes and that are in production today, already leak this sort of information (individual lives in a particular state).

Reply:
But do they otherwise leak it without getting any consent from (or even indication to) the user?

Reply:
> But do they otherwise leak it without getting any consent from (or even indication to) the user?

I was going off of this statement:

> applications registering to handle custom URL schemes effectively make their package name visible to other applications.

If that's true, then the current state-by-state mDL apps that register to handle the custom URL schemes (like `openid4vp://` as [spec'd here](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-10.1.2)) would fall into that category. I believe the attack that's being discussed here is:

1. US state mDL app registers to handle `openid4vp://`.
2. Attacker app gets all package names capable of handling `openid4vp://`.
3. Attacker now knows that the victim is most likely a resident of a specific US state.

That said, I don't what avenue one would use to do that on Android (I'm not a mobile phone developer), so it may be that understanding which app registered doesn't happen until the individual clicks on the app in the protocol scheme handler app selection sheet?

What I was trying to provide here was a concrete example of this happening today. The text makes it sound like registering to handle a custom URL scheme on Android bypasses the "keep list of installed applications private from other applications" protection. If so, it would help to provide a concrete example, if not, you might want to reword the above as it makes it seem like it's possible to bypass that protection today.

Reply:
Another point perhaps worth mentioning:

A browser API could potentially keep private (from the relying verifier / issuer) the specific wallet the user selects to complete a request in ways that may be impossible or challenging when using other mechanisms (e.g., a custom URL scheme that includes an invitation to use a protocol that requires a wallet to communicate as a client with the relying verifier / issuer).

Reply:
> I was going off of this statement:

Ah sorry, my bad.

>  If so, it would help to provide a concrete example (state-issued mDL apps), if not, you might want to reword the above as it makes it seem like it's possible to bypass that protection today.

I see. I don't actually know if any apps in production are registering for the mdoc: scheme today. We're monitoring mdoc: usage in Chrome and it's currently so small that I assume it's just a few developers doing testing. I've expanded on this and added a link for more details. It's a little complex, but my understanding is that yes registering to handle any intents (including custom URL schemes) makes your package name discoverable by other apps.

> A browser API could potentially keep private (from the relying verifier / issuer) the specific wallet the user selects to complete a request. This may be impossible or challenging when using other mechanisms (e.g., a custom URL scheme that includes an invitation to use a protocol that requires a wallet to directly communicate as a client with the relying verifier / issuer).

Yeah I was wondering about saying something like that, but I didn't see how it would be materially different. Both setups involve some arbitrary data transfer from the wallet back to the verifier. That protocol could be designed to either rely on wallet keys (or other form of wallet identification) or not, right?

Reply:
> my understanding is that yes registering to handle any intents (including custom URL schemes) makes your package name discoverable by other apps.

Right, and if that's the case, let's give people a concrete example: State-provided mDL apps identify that the individual resides in a particular state w/o the individual giving consent to releasing that information. This is something that CHAPI does not expose to the Verifier until after consent is provided by the Holder to release the information requested by the Verifier.

> > A browser API could potentially keep private (from the relying verifier / issuer) the specific wallet the user selects to complete a request. This may be impossible or challenging when using other mechanisms (e.g., a custom URL scheme that includes an invitation to use a protocol that requires a wallet to directly communicate as a client with the relying verifier / issuer).
> 
> Yeah I was wondering about saying something like that, but I didn't see how it would be materially different. Both setups involve some arbitrary data transfer from the wallet back to the verifier. 

The point I was trying to make was that a browser API could keep which wallets are available/selected hidden from the Verifier (and that would probably be ideal, even after consent to share a digital credential from the Holder). 

> That protocol could be designed to either rely on wallet keys (or other form of wallet identification) or not, right?

Yes, and I'll suggest that /one/ of the protocols is to just deliver the Presentation to the Verifier through the browser API w/o exposing anything about the wallet that performed the task. This optimizes for privacy in the default case. That is not to say that the Verifier couldn't also ask: "I want an attestation on which wallet you're using", but that has a bad "market competition" smell. There are some entities that believe that wallets must always identify themselves, which some in the ecosystem believe is an anti-pattern. The alternative is to perform capability checking on the wallet. In an ideal world, the Verifier should be asking:

* Provide a digital credential to me that verifies that the cryptographic key you are using to digitally sign the presentation is housed in an HSM.
* Provide a liveness credential that verifies that your face matches the face on the identity document you are supplying.
* Provide a captcha credential that proves that you solved this captcha I've provided that is specific to this credential exchange.

and not: "Provide a credential for the BigBrand wallet you are using so I can determine if I want to interact with your wallet because I have vetted it for usage for my use case." (Doing "wallet sniffing" feels quite similar to "browser sniffing", which we've tried desperately to move away from over the past two decades on the Web Platform).

In short, wallet identification feels like an anti-pattern and we should instead follow the path that the browsers have taken wrt. user agent capability detection.